src/django/api/views/boundary.py

```diff
@@ -0,0 +1,69 @@
+from django.http import Http404
+from django.db.models import Prefetch
+
+from rest_framework.response import Response
+from rest_framework.exceptions import APIException
+from rest_framework.views import APIView
+from rest_framework.permissions import IsAuthenticated
+
+from ..models.user import Roles
+from ..models.boundary import Boundary
+from ..models.utility import Utility
+from ..models.submission import Submission
+
+from ..serializers.boundary import BoundaryDetailSerializer, BoundaryListSerializer
+
+
+def get_boundary_queryset_for_user(user):
+    if user.role == Roles.CONTRIBUTOR:
+        return Boundary.objects.filter(utility__in=user.utilities.all())
+
+    if user.role == Roles.VALIDATOR:
+        return Boundary.objects.filter(utility__state__in=user.states.all())
+
+    if user.role == Roles.ADMINISTRATOR:
+        return Boundary.objects.all()
+
+    raise RuntimeError('Invalid role: {}'.format(user.role))
+
+
+class BoundaryListView(APIView):
+    permission_classes = [IsAuthenticated]
+
+    def get(self, request, format=None):
+        boundaries = get_boundary_queryset_for_user(request.user)
+        boundaries = boundaries.select_related('utility')
+
+        requested_utilities = request.GET.get('utilities')
+        if requested_utilities:
+            for utility_id in requested_utilities:
+                if not request.user.has_access_to_utility(
+                    Utility.objects.get(pk=utility_id)
+                ):
+                    raise APIException('Cannot access this utility', 401)
+
+            boundaries = boundaries.filter(utility__in=requested_utilities)
+
+        return Response(BoundaryListSerializer(boundaries, many=True).data)
+
+
+class BoundaryDetailView(APIView):
+    permission_classes = [IsAuthenticated]
+
+    def get(self, request, id, format=None):
+        boundary = get_boundary_queryset_for_user(request.user)
+        boundary = boundary.select_related('utility')
+        boundary = boundary.prefetch_related(
+            Prefetch(
+                'submissions',
+                queryset=Submission.objects.select_related('review').prefetch_related(
+                    'review__annotations'
+                ),
+            )
+        )
+
+        boundary = boundary.filter(pk=id).first()
+        if boundary is None:
+            raise Http404
```

Comment:
I think this would work? `boundary` is morphing from a queryset into an object, so you might consider a new name.
```suggestion
        boundary = get_object_or_404(boundary, pk=id)
```

Reply:
It does: ae3c80f78b5e885dde47dc4d58eb00496f457492